cloud-controller-manager/do/kubernetes_lb_validator.go

```diff
@@ -0,0 +1,165 @@
+		{
+			EntryProtocol:  "http",
+			EntryPort:      80,
+			TargetProtocol: "http",
+			TargetPort:     80,
+			CertificateID:  "",
+			TlsPassthrough: false,
+		}}
+
+	lbRequest := v.buildRequest(svc.Name, v.Region, v.DropletTag, forwardingRules)
+
+	var resp *godo.Response
+
+	if req.OldObject.Object == nil {
+		if len(svc.Status.LoadBalancer.Ingress) == 0 || svc.Status.LoadBalancer.Ingress[0].IP == "" {
+			resp, err = v.validateCreate(ctx, lbRequest)
+			if err != nil {
+				if v.checkError(resp) {
+					v.Log.Error(err, "failed to validate lb creation")
+					return admission.Denied(fmt.Sprintf("failed to validate lb creation: %v", err))
+				}
+				v.Log.Error(err, "failed to validate lb creation")
+				return admission.Errored(int32(resp.StatusCode), err)
+			}
+			return admission.Allowed("valid lb create request")
+		}
+	}
+
+	err = v.decoder.DecodeRaw(req.OldObject, svc)
+	if err != nil {
+		v.Log.Error(err, "failed to decode existing object")
+		return admission.Errored(http.StatusBadRequest, fmt.Errorf("failed to decode request: %v", err))
+	}
+
+	currentLBID := svc.Annotations["kubernetes.digitalocean.com/load-balancer-id"]
+	if currentLBID == "" {
+		v.Log.Error(fmt.Errorf("missing loadbalancer id in annotation"), "missing kubernetes.digitalocean.com/load-balancer-id annotation")
+		resp, err = v.validateCreate(ctx, lbRequest)
+		if err != nil {
+			if v.checkError(resp) {
+				v.Log.Error(err, "failed to validate lb creation")
+				return admission.Denied(fmt.Sprintf("failed to validate lb creation: %v", err))
+			}
+			v.Log.Error(err, "failed to validate lb creation")
+			return admission.Errored(int32(resp.StatusCode), err)
+		}
+		return admission.Allowed("can not validate lb without annotation: kubernetes.digitalocean.com/load-balancer-id annotation")
+	}
+
+	resp, err = v.validateUpdate(ctx, currentLBID, lbRequest)
+	if err != nil {
+		if v.checkError(resp) {
+			v.Log.Error(err, "failed to validate lb update")
+			return admission.Denied(fmt.Sprintf("failed to validate lb update: %v", err))
+		}
+		v.Log.Error(err, "failed to validate lb update")
+		return admission.Errored(int32(resp.StatusCode), err)
+	}
+	return admission.Allowed("valid update request")
+}
+
+func (v *KubernetesLBServiceValidator) validateCreate(ctx context.Context, lbRequest *godo.LoadBalancerRequest) (*godo.Response, error) {
+	_, resp, err := v.GClient.LoadBalancers.Create(ctx, lbRequest)
+	return resp, err
+}
+
+func (v *KubernetesLBServiceValidator) validateUpdate(ctx context.Context, currentLBID string, lbRequest *godo.LoadBalancerRequest) (*godo.Response, error) {
+	_, resp, err := v.GClient.LoadBalancers.Update(ctx, currentLBID, lbRequest)
+	return resp, err
+}
+
+func (v *KubernetesLBServiceValidator) buildRequest(name string, region string, tag string, forwardingRules []godo.ForwardingRule) *godo.LoadBalancerRequest {
+	return &godo.LoadBalancerRequest{
+		Name:            name,
+		Tag:             tag,
```

Comment:
As clarified out of band, an LB validation request does not require passing a list of droplet IDs or a tag. This means we can rip out any logic related to that and focus on validating the configuration only.